src/cost_estimates/fee_medians.rs

```diff
@@ -0,0 +1,348 @@
+use std::cmp;
+use std::cmp::Ordering;
+use std::convert::TryFrom;
+use std::{iter::FromIterator, path::Path};
+
+use rusqlite::AndThenRows;
+use rusqlite::Transaction as SqlTransaction;
+use rusqlite::{
+    types::{FromSql, FromSqlError},
+    Connection, Error as SqliteError, OptionalExtension, ToSql,
+};
+use serde_json::Value as JsonValue;
+
+use chainstate::stacks::TransactionPayload;
+use util::db::sqlite_open;
+use util::db::tx_begin_immediate_sqlite;
+use util::db::u64_to_sql;
+
+use vm::costs::ExecutionCost;
+
+use chainstate::stacks::db::StacksEpochReceipt;
+use chainstate::stacks::events::TransactionOrigin;
+
+use crate::util::db::sql_pragma;
+use crate::util::db::table_exists;
+
+use super::metrics::CostMetric;
+use super::FeeRateEstimate;
+use super::{EstimatorError, FeeEstimator};
+
+use super::metrics::PROPORTION_RESOLUTION;
+use cost_estimates::StacksTransactionReceipt;
+
+const CREATE_TABLE: &'static str = "
+CREATE TABLE median_fee_estimator (
+    measure_key INTEGER PRIMARY KEY AUTOINCREMENT,
+    high NUMBER NOT NULL,
+    middle NUMBER NOT NULL,
+    low NUMBER NOT NULL
+)";
+
+/// FeeRateEstimator with the following properties:
+///
+/// 1) We use a "weighted" percentile approach for calculating the percentile values. Described
+///    below, larger transactions contribute more to the ranking than small transactions.
+/// 2) Use "windowed" decay instead of exponential decay. This allows outliers to be forgotten
+///    faster, and so reduces the influence of outliers.
+/// 3) "Pad" the block, so that any unused spaces is considered to have an associated fee rate of
+///    1f, the minimum. Ignoring the amount of empty space leads to over-estimates because it
+///    ignores the fact that there was still space in the block.
+pub struct WeightedMedianFeeRateEstimator<M: CostMetric> {
+    db: Connection,
+    /// We only look back `window_size` fee rates when averaging past estimates.
+    window_size: u32,
+    /// The weight of a "full block" in abstract scalar cost units. This is the weight of
+    /// a block that is filled on each dimension.
+    full_block_weight: u64,
```

Comment:
I think that the "fill with empty" behavior is a little more honest if the "full" block weight is  a block with just _one_ dimension filled. Because a block cannot exceed any dimension, a block with a weight > a single dimension fill may not be able to accept any more transactions. Currently, I think this only impacts the docstring comment, but when it's integrated into the `testnet/stacks-node`, that will impact how this value is set.

Reply:
changed the text to say its any dimension